### ISSUE TITLE
sql: Error messages should be improved in case of no matching data source

### ISSUE THREAD

**Issue by rytaft** [cockroachdb/cockroach#22966]
Consider the following invalid query:
```
root@:26257/> select count(tpch.nation.n_nationkey) from tpch.customer c, tpch.public.nation n where tpch.public.customer.c_nationkey = nation.n_nationkey group by tpch.public.nation.n_nationkey limit 10;
```

The error that CockroachDB returns for this query is:
```
pq: no data source matches prefix: tpch.public.customer
```

PostgreSQL has a friendlier error message with a hint:
```
Postgres: ERROR: invalid reference to FROM-clause entry for table "nation" Hint: Perhaps you meant to reference the table alias "n". Position: 14
```

**Comment by knz** [comment]
I'll look what I can do. Note that I never like pg's error message that refers to a "FROM" clause, because it can also appear when there is no FROM clause whatsoever.

Meanwhile I'll just ignore pg's hint for now, since it's complex to implement we're not going to keep the current lookup code anyway.

But Becca I do have a hint for you: the object `pgerror.Error` has a `SetHint` method. I encourage you to implement any hint you'd like in the lookup code you recently implemented.

**Comment by rytaft** [comment]
Sounds good - thanks!

**Comment by knz** [comment]
Fixed in #27396.
